### Exercici 1
- Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció

In [7]:
import pandas as pd
import numpy as np

In [60]:
avions=pd.read_csv('DelayedFlights.csv')
avions.drop_duplicates(inplace = True)

avions.head()

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0           0  2008      1           3          4   2003.0        1955   
1           1  2008      1           3          4    754.0         735   
2           2  2008      1           3          4    628.0         620   
3           4  2008      1           3          4   1829.0        1755   
4           5  2008      1           3          4   1940.0        1915   

   ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0   2211.0        2225            WN  ...     4.0     8.0          0   
1   1002.0        1000            WN  ...     5.0    10.0          0   
2    804.0         750            WN  ...     3.0    17.0          0   
3   1959.0        1925            WN  ...     3.0    10.0          0   
4   2121.0        2110            WN  ...     4.0    10.0          0   

   CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                 N         0           NaN           NaN      NaN   
1                 N         0           NaN           NaN      NaN   
2                 N         0           NaN           NaN      NaN   
3                 N         0           2.0           0.0      0.0   
4                 N         0           NaN           NaN      NaN   

  SecurityDelay  LateAircraftDelay  
0           NaN                NaN  
1           NaN                NaN  
2           NaN                NaN  
3           0.0               32.0  
4           NaN                NaN  

[5 rows x 30 columns]

In [4]:
#Para saber los nulos segun columnas, lo mismo que utilizar df.isna().sum()
avions.isnull().sum()


Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Aquí vemos que las columnas de 'CarrierDelay', 'WeatherDelay', 'NasDelay', SecurityDelay, LateAircraftDelay, contienen el mismo número de valores nulos. Por lo que seran susceptibles de ser examinadas con detenimiento.

Según nos indican los datos del dataframe, los vuelos de -15min no se consideran vuelos retrasados, por lo tanto vamos a comprobar cuantos vuelos llegaron tarde -15minutos.

In [35]:
avions[avions['ArrDelay'] < 15].shape[0]

70735

Siguen sin salir los mismos números, por lo que pasaremos a comprovar el número de vuelos desviados a otro lugar así como el número de vuelos cancelados.

In [11]:
avions['Diverted'].value_counts() #Contar los vuelos desviados

Diverted
0    1929004
1       7754
Name: count, dtype: int64

In [12]:
avions['Cancelled'].value_counts() #Contar los vuelos cancelados

Cancelled
0    1936125
1        633
Name: count, dtype: int64

In [21]:
total = avions[avions['ArrDelay'] < 15].shape[0] + avions['Diverted'].sum() + avions['Cancelled'].sum()
total

689270

Todos aquellos valores nulos que aparecen en las últimas columnas son los mismos vuelos que fueron desviados, cancelados o no son considerados retrasados, porque el retraso no superó los 15min.
Por lo que estos calores los cambiaremos de nulos a 0.

In [61]:
avions[['CarrierDelay','WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']] = avions[['CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']].fillna(0)


In [62]:
#Buscamos los valores nulos
avions.isnull().sum()

Unnamed: 0              0
Year                    0
Month                   0
DayofMonth              0
DayOfWeek               0
DepTime                 0
CRSDepTime              0
ArrTime              7110
CRSArrTime              0
UniqueCarrier           0
FlightNum               0
TailNum                 5
ActualElapsedTime    8387
CRSElapsedTime        198
AirTime              8387
ArrDelay             8387
DepDelay                0
Origin                  0
Dest                    0
Distance                0
TaxiIn               7110
TaxiOut               455
Cancelled               0
CancellationCode        0
Diverted                0
CarrierDelay            0
WeatherDelay            0
NASDelay                0
SecurityDelay           0
LateAircraftDelay       0
dtype: int64

In [63]:
avions.shape

(1936758, 30)

Teniendo en cuenta el número de registros del dataset, y pensando que se realizarà una selección aleatoria a posteriori para poder trabajar mejor.
Se considera que se pueden eliminar los datos nulos que aparecen.

In [64]:
avions=avions.dropna()

In [65]:
#Buscamos los valores nulos
avions.isnull().sum()

Unnamed: 0           0
Year                 0
Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
FlightNum            0
TailNum              0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
TaxiIn               0
TaxiOut              0
Cancelled            0
CancellationCode     0
Diverted             0
CarrierDelay         0
WeatherDelay         0
NASDelay             0
SecurityDelay        0
LateAircraftDelay    0
dtype: int64

Aquí ya vemos como no tenemos valores nulos. 

In [66]:
#aleatoriament se recogen 200.000 registros
avions=avions.sample(n=200000)
avions.info()

<class 'pandas.core.frame.DataFrame'>
Index: 200000 entries, 1779631 to 1816289
Data columns (total 30 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         200000 non-null  int64  
 1   Year               200000 non-null  int64  
 2   Month              200000 non-null  int64  
 3   DayofMonth         200000 non-null  int64  
 4   DayOfWeek          200000 non-null  int64  
 5   DepTime            200000 non-null  float64
 6   CRSDepTime         200000 non-null  int64  
 7   ArrTime            200000 non-null  float64
 8   CRSArrTime         200000 non-null  int64  
 9   UniqueCarrier      200000 non-null  object 
 10  FlightNum          200000 non-null  int64  
 11  TailNum            200000 non-null  object 
 12  ActualElapsedTime  200000 non-null  float64
 13  CRSElapsedTime     200000 non-null  float64
 14  AirTime            200000 non-null  float64
 15  ArrDelay           200000 non-null  float64
 16  

- index: es igual que unnamed. 
    - se puede eliminar
- _unnamed: es un número de vol 
    - se puede eliminar
- _year: año, tots els vols son del 2008 
    - datos de tiempo. se puede eliminar
- _month:mes 
    - datos de tiempo. se puede eliminar
- _dayofmonth: dia del mes 
    - datos de tiempo. se puede eliminar
- _DepTime: Departure Time o Hora de salida del avión 
    - datos de tiempo. se puede eliminar
- CRSDepTime: Scheduled Departure Time o Hora de Llegada Programada
    - datos de tiempo. se puede eliminar
- ArrTime: Arrival Time o Hora de llegada del avión
- CRSArrTime: Scheduled Arrival Time o Hora de Llegada Programada
- UniqueCarrier: Código de identificación único de una aerolínea para registrar e identificar aerolíneas.
- FlightNum: Número de vuelo
- TailNum: Número de
- ActualElapsedTime: ActualElapsedTime. Tiempo real que ha transcurrido desde el despegue hasta el aterrizaje de un vuelo específico. 
- CRSElapsedTime: Scheduled Elapsed Time. Tiempo estimado de vuelo antes de que despegue sin retrasos.
- AirTime: Tiempo en el aire. Tiempo de vuelo 
- ArrDelay: Arrival Delay. Tiempo que el vuelo ha llegado con retraso.
- TaxiIn: Rodaje de Llegada. Tiempo que una aeronave pasa rodando en la pista después de aterrizar y antes de llegar a la puerta de llegada en el aeropuerto de destino. 
    - se puede eliminar no interesa para nuestros datos 
- _TaxiOut: Rodaje de Salida. Tiempo que una aeronave pasa rodando en la pista antes de despegar.
    - se puede eliminar no interesa para nuestros datos 
- Cancelled: Cancelado. Indica si un vuelo ha sido cancelado o no.
- CancellationCode: Código de Cancelación Proporciona una razón específica para la cancelación de un vuelo.
    - dara Nan
- Diverted:Desviado. Indica si un vuelo ha sido desviado a un aeropuerto diferente al de destino originalmente planeado. Puede tener un valor binario, 1 para indicar que el vuelo fue desviado y 0 para indicar que no fue desviado
- CarrierDelay: El retraso atribuido a la aerolínea, como problemas logísticos o de programación. 
    - Se puede eliminar hay muchos valores Nan
- WeatherDelay: El retraso causado por condiciones climáticas adversas.
    - Se puede eliminar hay muchos valores Nan
- NASDelay: National Airspace System Delay o Sistema espacio aéreo nacional. Retraso relacionado con la congestión o problemas en el sistema de espacio aéreo nacional.
    - Se puede eliminar hay muchos valores Nan
- SecurityDelay:  Retraso causado por cuestiones de seguridad.
    - Se puede eliminar hay muchos valores Nan
- LateAircraftDelay: Retraso causado por la llegada tardía de la aeronave previa.
 
En aquest cas, les dades en relació a la data de l'avió o la hora concreta no ens interessen i seran complicades de treballar en relació amb les altres pel que no es tindran en compte. Tampoc interessa el temps que l'avió està a pista, el motiu pel qual s'ha retrassat o si no ha arribat a enlairar.se.
Amb tota la informació que tenim del dataset, agafarem un número reduit de columnes per poder treballar millor.

In [67]:
avions=avions[['UniqueCarrier','FlightNum', 'TailNum', 'AirTime', 'ArrDelay','DepDelay', 'Origin', 'Dest', 'Distance']]

avions

UniqueCarrier  FlightNum TailNum  AirTime  ArrDelay  DepDelay Origin  \
1779631            XE       2732  N12157    118.0       4.0       7.0    IAH   
1687634            EV       5383  N752EV    113.0      82.0      82.0    ATL   
437541             OO       3993  N426SW     72.0      39.0      44.0    DEN   
1344899            WN       2012  N908WN    139.0      -4.0       7.0    BUF   
1398424            AA       1768  N237AA     92.0       6.0      28.0    ORD   
...               ...        ...     ...      ...       ...       ...    ...   
1486901            FL        787  N935AT    128.0      79.0      31.0    ATL   
128802             MQ       4561  N717AE    136.0      36.0      44.0    RDU   
1073960            CO        648  N12216    348.0      51.0      74.0    EWR   
1808100            OO       6620  N713SK     24.0     229.0     231.0    ASE   
1816289            UA        688  N853UA    107.0     149.0     133.0    ORD   

        Dest  Distance  
1779631  CLT       913  
1687634  OKC       761  
437541   SLC       391  
1344899  TPA      1054  
1398424  PHL       678  
...      ...       ...  
1486901  PHL       665  
128802   XNA       868  
1073960  SFO      2565  
1808100  DEN       125  
1816289  LGA       733  

[200000 rows x 9 columns]

Se coge un dataframe reducido, solo con aquellas columnas que nos interessan.

In [68]:
#Resumen de los datos
avions.info()

<class 'pandas.core.frame.DataFrame'>
Index: 200000 entries, 1779631 to 1816289
Data columns (total 9 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   UniqueCarrier  200000 non-null  object 
 1   FlightNum      200000 non-null  int64  
 2   TailNum        200000 non-null  object 
 3   AirTime        200000 non-null  float64
 4   ArrDelay       200000 non-null  float64
 5   DepDelay       200000 non-null  float64
 6   Origin         200000 non-null  object 
 7   Dest           200000 non-null  object 
 8   Distance       200000 non-null  int64  
dtypes: float64(3), int64(2), object(4)
memory usage: 15.3+ MB


Vemos el tipo de variables del dataset, valores sin registros nulos.

### Exercici 2
- Fes un informe complet del dataset:

Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.
Troba quantes dades faltants hi ha per columna.
Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...).
Fes una taula de les aerolínies amb més endarreriments acumulats.
Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards.
Aporta allò que consideris rellevant.

In [70]:
avions_analisis= avions[['Distance','AirTime','ArrDelay', 'DepDelay' ]]
resumen = round(avions_analisis.describe(),2)
resumen

Distance    AirTime   ArrDelay   DepDelay
count  200000.00  200000.00  200000.00  200000.00
mean      765.01     108.28      42.32      43.18
std       573.85      68.59      56.95      53.46
min        31.00       2.00     -66.00       6.00
25%       338.00      58.00       9.00      12.00
50%       606.00      90.00      24.00      24.00
75%       998.00     137.00      56.00      53.00
max      4962.00     634.00    1707.00    1710.00

En el cuadro aparece el resumen estadístico de las variables numéricas que nos interesa observar. En este caso,  la distancia, el tiempo de vuelo y el tiempo de los aviones retrasados tanto de llegada como de salida.
Se observa como la distancia media es 765 millas.
El tiempo de vuelo tiene una mediana de 108 minutos y la distancia de 765 millas.
En relación a las medias de las salidas y las llegadas, son muy parecidas entre 42 y 43 min. Por lo que los vuelos tienden a salir tarde y llegar tarde. Si tenemos en cuenta las medias, podemos hacer la hipotesis de que no acortan en tiempo de vuelo el retraso de la salida.

-  Troba quantes dades faltants hi ha per columna.

In [71]:
valors_nuls_por_columna = avions.isna().sum()

print(valors_nuls_por_columna)

UniqueCarrier    0
FlightNum        0
TailNum          0
AirTime          0
ArrDelay         0
DepDelay         0
Origin           0
Dest             0
Distance         0
dtype: int64


No quedan datos nulos en el dataframe con el que estamos trabajando.

#### - Crea columnes noves (velocitat mitjana del vol, si ha arribat tard o no...). 

In [30]:
#Calculo de la velocitat mitjana
avions['MedSpeed']=round(avions['Distance']/round(avions['AirTime']/60,3).mean(),2)
avions[['Distance','MedSpeed']]

Distance  MedSpeed
1482252       239    132.35
1843610      1968   1089.84
232708        441    244.22
110171       1199    663.98
420077        657    363.83
...           ...       ...
1495252       865    479.02
425106       1297    718.25
1884546       487    269.69
447242        626    346.67
1353107       134     74.21

[200000 rows x 2 columns]

La velocidad está en milles per hora. Aquí el AirTime en minuts.
En esta tabla se muestra el número de vuelo, la distancia en millas y la media de millas por hora.
Se crea una nueva columna que se llama MedSpeed para la distancia media de los aviones.

In [31]:
#numpy.where(condició, valor_si_vertader, valor_si_fals)
avions['ArrDelay'].sort_values(ascending=False)
avions['Delay']=np.where(avions['ArrDelay']>15,'Delayed','OnTime')
avions[['ArrDelay','Delay']]


ArrDelay    Delay
1482252      20.0  Delayed
1843610     161.0  Delayed
232708       87.0  Delayed
110171        4.0   OnTime
420077      160.0  Delayed
...           ...      ...
1495252      10.0   OnTime
425106       11.0   OnTime
1884546      48.0  Delayed
447242       22.0  Delayed
1353107       5.0   OnTime

[200000 rows x 2 columns]

Se crea una columna nueva llamada Delay. Para saber si ha llegado tarde o no un vuelo se crea una columna de tipo object. Para mostrar los strings "Delayed" o "OnTime". Se debe tener en cuenta que solo si ha llegado más de 15 minutos tarde se considera que un vuelo se ha retrasado. Por lo que se creará una columna nueva donde conste si el vuelo se ha retrasado o no. Luego se puede ver en la tabla, si se observa la columna de ArrDelay junto con la de Delay. Funciona correctamente.

  #### - Fes una taula de les aerolínies amb més endarreriments acumulats.

In [94]:
avions.groupby(['UniqueCarrier'])['ArrDelay'].sum().sort_values(ascending=False)

UniqueCarrier
WN    1175966.0
AA     912033.0
UA     700874.0
MQ     669924.0
OO     604393.0
XE     547749.0
DL     470924.0
CO     432045.0
EV     398999.0
YV     382814.0
US     369544.0
NW     369434.0
FL     317884.0
B6     316373.0
OH     280797.0
9E     251633.0
AS     150257.0
F9      84426.0
HA      26072.0
AQ       1045.0
Name: ArrDelay, dtype: float64

Aquí se crea una tabla nueva con aquellas aerolíneas que acumulan más retrasos. Se calcula la suma del retraso de llegada para cada aerolínea.
Primero se agrupan las aerolínias según el retraso de llegada de vuelos. Y se ordenan de manera descendente. Por lo que la primera aerolínea que aparece serà la que tiene el mayor retraso en minutos acumulados por cada vuelo realizado.

#### - Quins són els vols més llargs? I els més endarrerits? Busca les rutes més llargues i les que acumulen més retards. Aporta allò que consideris rellevant.

In [72]:
#vols més llargs

round(avions.groupby(['FlightNum','Origin','Dest'])['AirTime'].mean()).sort_values(ascending=False)

FlightNum  Origin  Dest
15         EWR     HNL     604.0
851        ATL     HNL     558.0
1273       ATL     HNL     554.0
14         HNL     EWR     542.0
1561       ATL     HNL     540.0
                           ...  
6277       CVG     LEX      12.0
5086       CVG     LEX      11.0
2800       MKE     GRB      11.0
5171       RSW     MIA       4.0
3703       SLC     SUN       2.0
Name: AirTime, Length: 56372, dtype: float64

Para saber los vuelos más largos se tiene en cuenta la media del tiempo que el vuelo ha estado en el aire. Luego ordenamos los vuelos de forma descendente para observar cuales fueron los que estuvieron de media más tiempo en el aire, y vemos el origen y el destino.

In [98]:
#las rutas más largas

round(avions.groupby(['FlightNum','Origin','Dest'])['Distance'].mean()).sort_values(ascending=False)


FlightNum  Origin  Dest
14         HNL     EWR     4962.0
15         EWR     HNL     4962.0
1560       HNL     ATL     4502.0
851        ATL     HNL     4502.0
850        HNL     ATL     4502.0
                            ...  
71         JNU     GST       41.0
77         GST     JNU       41.0
5431       SNA     LAX       36.0
64         PSG     WRG       31.0
65         WRG     PSG       31.0
Name: Distance, Length: 56372, dtype: float64

Miramos las rutas más largas teniendo en cuenta la media de la distancia de vuelo. Luego ordenamos los vuelos de forma descendente para observar cuales fueron los que realizaron de media la ruta más larga,además de su origen y su destino.

In [99]:
#las rutas que acumulan más retrasos

round(avions.groupby(['FlightNum','Origin','Dest'])['ArrDelay'].sum(),2).sort_values(ascending=False)


FlightNum  Origin  Dest
534        TPA     EWR     3340.0
1555       MIA     SJU     3126.0
7230       ORD     CAE     2813.0
1076       RIC     JFK     2679.0
2994       CLT     EWR     2528.0
                            ...  
27         PHL     STL      -69.0
504        STT     JFK      -72.0
1981       PHX     PVD      -73.0
483        SNA     OGG      -85.0
467        SMF     OGG      -85.0
Name: ArrDelay, Length: 56372, dtype: float64

Miramos las rutas que acumulan más retrasos teniendo en cuenta la suma de arrival delay, que seria la columna que nos indica aquellos vuelos que han llegado tarde. Luego ordenamos de manera descendente, para que queden encima los vuelos más retrasados.
Vemos que la ruta con más retrasos es la de Tampa a Newark.

  ## Exercici 3

- Exporta el dataset net i amb les noves columnes a Excel.

In [100]:
avions.to_excel('Excel_avions.xlsx', index=False)